featurebyte/service/deploy.py

```diff
@@ -288,221 +368,395 @@ async def _get_enabled_serving_entity_ids(
+            )
+
+        # update deployment status
+        await self.deployment_service.update_document(
+            document_id=deployment.id,
+            data=DeploymentUpdate(enabled=True),
+            return_document=False,
+        )
+        await self._update_progress(100, "Deployment enabled successfully!")
+
+    async def _check_feature_list_used_in_other_deployment(
+        self, feature_list_id: ObjectId, exclude_deployment_id: ObjectId
+    ) -> bool:
+        # check whether other deployment are using this feature list
+        list_deployment_results = await self.deployment_service.list_documents_as_dict(
+            query_filter={
+                "feature_list_id": feature_list_id,
+                "enabled": True,
+                "_id": {"$ne": exclude_deployment_id},
+            }
+        )
+        return bool(list_deployment_results["total"] > 0)
+
+    async def enable_deployment(
+        self,
+        deployment: DeploymentModel,
+        feature_list: FeatureListModel,
+    ) -> None:
+        """
+        Enable deployment
+
+        Parameters
+        ----------
+        deployment: DeploymentModel
+            Target deployment
+        feature_list: FeatureListModel
+            Feature list of the deployment
+
+        Raises
+        ------
+        DocumentUpdateError
+            When there is an unexpected error during deployment enablement
+        """
+        try:
+            await self._enable_deployment(deployment, feature_list)
+        except Exception as exc:
+            exc_traceback = traceback.format_exc()
+            logger.error(
+                f"Failed to enable deployment {deployment.id}: {exc}. {exc_traceback}",
+                exc_info=True,
+            )
+            await self.disable_deployment(deployment, feature_list)
+            raise DocumentUpdateError("Failed to enable deployment") from exc
+
+    async def disable_deployment(
+        self,
+        deployment: DeploymentModel,
+        feature_list: FeatureListModel,
+    ) -> None:
+        """
+        Disable deployment
+
+        Parameters
+        ----------
+        deployment: DeploymentModel
+            Target deployment
+        feature_list: FeatureListModel
+            Feature list of the deployment
+        """
+        # enable features online
+        feature_update_progress = get_ranged_progress_callback(
+            self.task_progress_updater.update_progress, 0, 70
+        )
+        features = []
+        features_offline_feature_table_to_update = []
+        for feature_id in feature_list.feature_ids:
+            feature = await self.feature_management_service.feature_service.get_document(
+                document_id=feature_id
+            )
+            updated_feature = await self.feature_management_service.online_disable_feature(
+                feature=feature,
+                feature_list=feature_list,
+            )
```

Comment:
Calling `online_disable_feature()` doesn't necessarily make a feature online disabled, but more to remove the reference of the feature list from that feature. Is that correct?

Reply:
It doesn't necessarily disable a feature online. It is more like delegate the action to `feature_management_service` to perform further check.